### PR TITLE
[SYCLomatic] Migrate 4 Cooperative Groups APIs

### PR DESCRIPTION
**Pull request by zhiweij1**
cg::tiled_partition<SIZE>(thread_block) => dpct::logical_group(item, item.get_group(), SIZE)
cg::thread_block_tile::size() => dpct::logical_group::get_local_linear_range()
cg::thread_block_tile::thread_rank() => dpct::logical_group::get_local_linear_id()
cg::thread_block_tile::shfl_down() => dpct::shift_sub_group_left()

### CODE REVIEW

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 1]
```diff
@@ -346,6 +346,51 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
                                      : id);
 }
 
+/// The logical-group is the group of some work-items.
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]: Could you further define "some"? 
The logical-group is the group of some work-items from work-group.  The num of work-items can be.....  It diffs with sub-group on size...

**Review by zhimingwang36 — APPROVED** [review]

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 2]
```diff
@@ -346,6 +346,50 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
...
+  cl::sycl::group<3> _g;
+  uint32_t _logical_group_size;
+  uint32_t _group_linear_range_in_parent;
+public:
+  /// Dividing \p parent_group into several logical-groups.
+  /// \param [in] parent_group The group to be divided.
+  /// \param [in] size The logical-group size.
+  logical_group(cl::sycl::group<3> parent_group, uint32_t size)
+      : _g(parent_group), _logical_group_size(size) {
+    _group_linear_range_in_parent =
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]: we should do extra checker on the size. (<=sub-group-size)

File: clang/test/dpct/helper_files_ref/include/util.hpp (hunk 1 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/util.hpp]: done

File: clang/test/dpct/helper_files_ref/include/util.hpp (hunk 2 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/util.hpp]: We cannot throw exception in device code.
So I set the logical group size to 1 if it is > sub-group size and documented in the comment.

**Comment by zhiweij1** [comment]
The end to end test PR: https://github.com/oneapi-src/SYCLomatic-test/pull/18

File: clang/test/dpct/helper_files_ref/include/util.hpp [hunk 3]
```diff
@@ -466,6 +466,59 @@ nd_range_barrier(cl::sycl::nd_item<1> item,
 
   item.barrier();
 }
+
+/// The logical-group is a logical collection of some work-items within a
+/// work-group. Each logical-group size is a power of 2 in the range
+/// [1, current_sub_group_size]. If the logical-group size is illegal,
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/util.hpp]: it's caller's respond to make sure logical-group size is in the range,  and is a number =2**N.  (device code has not way to through exception here).

Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/util.hpp]: OK, I will remove the "if" block.

Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/util.hpp]: done

**Review by ziranzha — APPROVED** [review]

**Review by YuriPlyakhin — COMMENTED** [review]
LGTM as experimental feature

**Pull request merged**